### PR TITLE
feature(rest): port storage/internal curl types to internal for future use

### PR DESCRIPTION
**Pull request by scotthart**
These files have been copied over from storage/internal "functionally as is". Only include paths, namespaces, etc. have been modified. Eventually when GCS is refactored to use the new REST library the copies of these files in storage/internal will be removed.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7745)

### CODE REVIEW

**Review by coryan — CHANGES_REQUESTED** [review]
These files are not compiled anywhere, is that intentional?  If so, what is the next step?  Add them to what library?

And once you add them to a (new?) library: how do you avoid the duplication?

File: google/cloud/BUILD.bazel [hunk 1]
```diff
@@ -157,3 +157,30 @@ load(":google_cloud_cpp_grpc_utils_benchmarks.bzl", "google_cloud_cpp_grpc_utils
         "@com_google_googleapis//google/spanner/v1:spanner_cc_grpc",
     ],
 ) for test in google_cloud_cpp_grpc_utils_benchmarks]
+
+load(":google_cloud_cpp_rest_internal.bzl", "google_cloud_cpp_rest_internal_hdrs", "google_cloud_cpp_rest_internal_srcs")
+
+cc_library(
+    name = "google_cloud_cpp_rest_internal",
```
Comment (coryan) [inline google/cloud/BUILD.bazel:164]: nit: the default visibility here is "public", consider changing the visibility of this target to "packages" or something similar.

File: google/cloud/CMakeLists.txt [hunk 2]
```diff
@@ -606,3 +606,155 @@ if (GOOGLE_CLOUD_CPP_ENABLE_GRPC)
...
+        PROPERTIES EXPORT_NAME "google-cloud-cpp::rest-internal"
+                   VERSION ${PROJECT_VERSION} SOVERSION
+                                              ${PROJECT_VERSION_MAJOR})
+    add_library(google-cloud-cpp::rest-internal ALIAS
+                google_cloud_cpp_rest_internal)
+
+    include(CreateBazelConfig)
+    create_bazel_config(google_cloud_cpp_rest_internal YEAR 2021)
+
+    # Define the benchmarks in a function so we have a new scope for variable
```
Comment (coryan) [inline google/cloud/CMakeLists.txt:646]: nit: s/benchmarks/unit tests/ ?

File: google/cloud/internal/curl_options.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ * them.
+ *
+ * @warning The library may create more connections than this option configures,
+ * for example if your application requests many simultaneous downloads.
+ */
+struct ConnectionPoolSizeOption {
+  using Type = std::size_t;
+};
+
+/**
```
Comment (coryan) [inline google/cloud/internal/curl_options.h:72]: This looks to me like a storage-only option.

File: google/cloud/internal/curl_options.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * When using formatted I/O operations (typically `operator>>(std::istream&...)`
+ * this option controls the size of the in-memory buffer kept to satisfy any I/O
+ * requests.
+ *
+ * Applications seeking optional performance for downloads should avoid
+ * formatted I/O, and prefer using `std::istream::read()`. This option has no
+ * effect in that case.
+ */
+struct DownloadBufferSizeOption {
```
Comment (coryan) [inline google/cloud/internal/curl_options.h:83]: same here, this is specific to GCS.

File: google/cloud/internal/curl_options.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ * When using formatted I/O operations (typically `operator<<(std::istream&...)`
+ * this option controls the size of the in-memory buffer kept before a chunk is
+ * uploaded. Note that GCS only accepts chunks in multiples of 256KiB, so this
+ * option is always rounded up to the next such multiple.
+ *
+ * Applications seeking optional performance for downloads should avoid
+ * formatted I/O, and prefer using `std::istream::write()`. This option has no
+ * effect in that case.
+ */
+struct UploadBufferSizeOption {
```
Comment (coryan) [inline google/cloud/internal/curl_options.h:99]: And here.

File: google/cloud/internal/curl_options.h [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+ * Disables automatic OpenSSL locking.
+ *
+ * With older versions of OpenSSL any locking must be provided by locking
+ * callbacks in the application or intermediate libraries. The C++ client
+ * library automatically provides the locking callbacks. If your application
+ * already provides such callbacks, and you prefer to use them, set this option
+ * to `false`.
+ *
+ * @note This option is only useful for applications linking against
+ * OpenSSL 1.0.2.
```
Comment (coryan) [inline google/cloud/internal/curl_options.h:113]: I wonder if OpenSSL 1.0.2 is a thing we can ignore now?

Comment (coryan) [reply google/cloud/internal/curl_options.h:113]: Maybe just file a bug to follow up on this.

**Review by coryan — APPROVED** [review]

File: ci/etc/expected_install_directories [hunk 7]
```diff
@@ -53,5 +53,6 @@
 ./lib64/cmake/google_cloud_cpp_iam
 ./lib64/cmake/google_cloud_cpp_logging
 ./lib64/cmake/google_cloud_cpp_pubsublite
+./lib64/cmake/google_cloud_cpp_rest_internal
```
Comment (devbww) [inline ci/etc/expected_install_directories:56]: `generate-libraries` seems to remove this for me.

Comment (coryan) [reply ci/etc/expected_install_directories:56]: My naming is awful, as usual.  `ci/etc/expected_install_directories` only contains the directories created by automatically generated libraries.  `ci/cloudbuilds/builds/cmake-install.sh` contains the list of directories created by hand-crafted libraries.

Comment (coryan) [reply ci/etc/expected_install_directories:56]: Meaning: we should move that hand-crafted directory to the `cmake-install.sh` script.

**Pull request merged**